lib/pengine/rules.c

```diff
@@ -1025,6 +1025,36 @@ accept_attr_expr(const char *l_val, const char *r_val, const char *type,
     return false;   // Should never reach this point
 }
 
+/*!
+ * \internal
+ * \brief Get correct value according to value-source
+ *
+ * \param[in] value         value given in rule expression
+ * \param[in] value_source  value-source given in rule expressions
+ * \param[in] match_data    If not NULL, resource back-references and params
+ */
+static const char *
+expand_value_source(const char *value, const char *value_source,
+                    pe_match_data_t *match_data)
+{
+    GHashTable *table = NULL;
+
+    if (pcmk__str_eq(value_source, "param", pcmk__str_casei)) {
+        table = match_data->params;
+
+    } else if (pcmk__str_eq(value_source, "meta", pcmk__str_casei)) {
+        table = match_data->meta;
+
+    } else { // literal
+        return value;
+    }
+
+    if ((table == NULL) || pcmk__str_empty(value)) {
+        return NULL;
```

Comment:
with an empty value we could bail out early

Reply:
good point

Reply:
It would change the behavior slightly: if value is the empty string ("") and value-source is literal, the current code would return the empty string whereas the early check would return NULL. That would make a difference for "defined/"not_defined" checks and when comparing against another empty string (NULL < empty string).

It seems to me the early check makes more sense, since the behavior for empty string vs NULL shouldn't differ based on value-source. I'll add a separate commit for it.

Reply:
Actually that changed behavior and the simpler return-checking was my initial reason to bring it up.